File: 220902/JavaScript/claw-machine.js
// 문제 해결 과정
// 1. Java에서는 for문에서 배열을 종으로 참조하기 위해 인덱스 변수값을 계속해서 바꿔가며 사용했다.
//    JavaScript에서는 선언형 방식을 사용하기 위해 foreach를 이용해 배열을 횡으로 참조할 수 있도록
//    배열의 횡방향과 종방향을 바꾼다.
//
// 2. 이제 아래 과정을 moves의 모든 숫자들에 대해 순차적으로 수행한다.
//   2-1. moves에서 숫자를 하나 가져온다. 그 숫자는 board에서 살펴볼 행(row)
//   2-2. 행의 값을 순차적으로 하나씩 꺼내보는데 0이 아닌 숫자가 나오면
//      그 숫자를 뽑은 인형 List에 넣고, 찾은 위치에는 찾은 숫자 대신 0을 재할당
//   2-3. 뽑은 인형 List에서 방금 List에 넣은 숫자랑 앞 숫자가 같은지 검사한다. ([size - 2] ?== [size - 1])
//      같다? 방금 List에 넣은 숫자랑 앞 숫자를 제거하고, 제거한 인형 숫자 변수에 +2
//
// 3. 제거한 인형 숫자 변수를 반환한다.

// reduce 메서드 활용법 reference
// https://www.zerocho.com/category/JavaScript/post/5acafb05f24445001b8d796d
// https://velog.io/@dyongdi/JS-2%EC%B0%A8%EC%9B%90-%EB%B0%B0%EC%97%B4%EC%9D%98-%ED%96%89%EA%B3%BC-%EC%97%B4-%EB%B0%94%EA%BE%B8%EA%B8%B0-Transposing-a-2D-array-in-JavaScript

export function transpose(board) {
  return board.reduce((newBoard, row) => (
    row.map((_, index) => (
      [
        ...(newBoard[index] || []), row[index],
      ]
    ))
  ), []);
}

export function pick(boardTransposed, pickedThings, row) {
  const pickedThing = boardTransposed[row - 1].find((element) => element > 0);

  if (pickedThing === undefined) {
    return;
  }

  pickedThings.push(pickedThing);

  const indexOfPickedThing = boardTransposed[row - 1].findIndex((element) => element > 0);
  boardTransposed[row - 1][indexOfPickedThing] = 0;
}

export function checkPickedThings(pickedThings) {
  if (pickedThings.length > 1
    && pickedThings[pickedThings.length - 1] === pickedThings[pickedThings.length - 2]) {
    pickedThings.pop();
    pickedThings.pop();

    return 2;
  }

  return 0;
}

export function solution(board, moves) {
  const boardTransposed = transpose(board);

  const pickedThings = [];

  const removedCount = {
    count: 0,
  };

  moves.forEach((row) => {
    pick(boardTransposed, pickedThings, row);

    removedCount.count += checkPickedThings(pickedThings);
  });

  return removedCount.count;
}

File: 220902/Java/ClawMachine.java
// 문제 해결 과정
// 아래 과정을 moves의 모든 숫자들에 대해 순차적으로 수행한다.
//   1. moves에서 숫자를 하나 가져온다.
//   2. board[0~4][숫자 - 1]만큼 돌면서 0이 아닌 숫자를 찾으면  (숫자 - 1인 이유: 인덱스니까!)
//      그 숫자를 다른 List에 넣고 board에서 찾은 위치에는 그 숫자 대신 0을 재할당
//   3. List에서 앞엣놈랑 방금 List에 넣은 놈이랑 같은지 검사한다. ([size - 2] ?= [size - 1])
//      같다? 방금 List에 넣은 놈이랑 앞엣놈을 제거하고, 제거한 인형 숫자 변수에 + 2
// 제거한 인형 숫자 변수를 반환한다.

import java.util.ArrayList;
import java.util.List;

public class ClawMachine {
  public int solution(int[][] board, int[] moves) {
    int length = board.length;

    List<Integer> pickedThings = new ArrayList<>();

    int removedCount = 0;

    for (int column : moves) {
      for (int i = 0; i < length; i += 1) {
        int picked = board[i][column - 1];

        if (picked != 0) {
          pickedThings.add(picked);

          board[i][column - 1] = 0;

          break;
        }
      }

      int lastIndex = pickedThings.size() - 1;

      if (lastIndex >= 1 &&
          pickedThings.get(lastIndex).equals(pickedThings.get(lastIndex - 1))) {
        pickedThings.remove(lastIndex);
        lastIndex -= 1;
        pickedThings.remove(lastIndex);

        removedCount += 2;
      }
    }

    return removedCount;
  }
}
